Utilities/DataFlow/src/SubframeBuilderDevice.cxx

```diff
@@ -83,60 +108,33 @@ bool o2::DataFlow::SubframeBuilderDevice::BuildAndSendFrame(FairMQParts &inParts
+  dh.subSpecification = mFLPId;
   dh.payloadSize = sizeof(SubframeMetadata);
 
+  DataHeader payloadheader(*o2::Header::get<DataHeader>((byte*)inParts.At(0)->GetData()));
+
   // subframe meta information as payload
   SubframeMetadata md;
-  // md.startTime = (hbh->orbit / mOrbitsPerTimeframe) * mDuration;
-  md.startTime = static_cast<uint64_t>(hbh->orbit) * static_cast<uint64_t>(mOrbitDuration);
-  md.duration = mDuration;
-  LOG(INFO) << "Start time for subframe (" << hbh->orbit << ", "
-                                           << mDuration
-            << ") " << timeframeIdFromTimestamp(md.startTime, mDuration) << " " << md.startTime<< "\n";
-
-  // send an empty subframe (no detector payload), only the data header
-  // and the subframe meta data are added to the sub timeframe
-  // TODO: this is going to be changed as soon as the device implements
-  // handling of the input data
+  // id is really the first orbit in the timeframe.
+  md.startTime = id.timeframeId * mOrbitsPerTimeframe * static_cast<uint64_t>(mOrbitDuration);
+  md.duration = mOrbitDuration*mOrbitsPerTimeframe;
+  LOG(INFO) << "Start time for subframe (" << md.startTime << ", "
+                                           << md.duration
+            << ")";
+
+  // Add the metadata about the merged subtimeframes
+  // FIXME: do we really need this?
   O2Message outgoing;
-
-  // build multipart message from header and payload
   AddMessage(outgoing, dh, NewSimpleMessage(md));
 
-  char *sourcePayload = nullptr;
-  auto payloadSize = extractDetectorPayload(&sourcePayload,
-                                            incomingBuffer,
-                                            inParts.At(1)->GetSize());
-  LOG(INFO) << "Got "  << inParts.Size() << " parts\n";
-  for (auto pi = 0; pi < inParts.Size(); ++pi)
-  {
-    LOG(INFO) << " Part " << pi << ": " << inParts.At(pi)->GetSize() << " bytes \n";
-  }
-  if (payloadSize <= 0)
-  {
-    LOG(ERROR) << "Payload is too small: " << payloadSize << "\n";
-    return true;
-  }
-  else
-  {
-    LOG(INFO) << "Payload of size " << payloadSize << "received\n";
-  }
-
-  auto *payload = new char[payloadSize]();
-  memcpy(payload, sourcePayload, payloadSize);
-  DataHeader payloadheader(*o2::Header::get<DataHeader>((byte*)inParts.At(0)->GetData()));
-
-  payloadheader.subSpecification = 0;
-  payloadheader.payloadSize = payloadSize;
-
-  // FIXME: take care of multiple HBF per SubtimeFrame
+  // Add the actual merged payload.
   AddMessage(outgoing, payloadheader,
-             NewMessage(payload, payloadSize,
-                        [](void* data, void* hint) { delete[] reinterpret_cast<char *>(hint); }, payload));
+             NewMessage(*outBuffer, outSize,
+                        [](void* data, void* hint) { delete[] reinterpret_cast<char *>(hint); }, *outBuffer));
   // send message
   Send(outgoing, mOutputChannelName.c_str());
+  // FIXME: do we actually need this? outgoing should go out of scope
```

Comment:
Al right, we can argue that `outgoing` is a local variable and should not be used after the Send.  The list will still contain the individual messages but they will have empty unique_ptr objects. Like it is now we do not need to clean because the local variable is destroyed anyhow.